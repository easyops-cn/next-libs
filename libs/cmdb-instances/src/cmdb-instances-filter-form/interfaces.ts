import { QueryOperatorValue } from "./constants/query";

export type QueryConditionValue = string | number | string[];

export interface QueryCondition {
  fieldId?: string;
  operator?: QueryOperatorValue;
  value?: QueryConditionValue;
}
