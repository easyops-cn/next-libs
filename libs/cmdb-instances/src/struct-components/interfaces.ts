import { ModelObjectAttrValue } from "@sdk/cmdb-sdk/dist/types/model/cmdb";

export interface Structkey {
  id: string;
  name: string;
  type: ModelObjectAttrValue["type"];
  regex?: any;
  protected?: boolean;
}
export interface AttributeValue {
  default?: any;
  default_type?: string;
  struct_define?: Structkey[];
  type: ModelObjectAttrValue["type"];
}
export interface Attribute {
  id?: string;
  name: string;
  protected?: boolean;
  readonly?: "true" | "false";
  required?: "true" | "false";
  tag?: string[];
  unique?: "true" | "false";
  value: AttributeValue;
  wordIndexDenined?: boolean;
  tips?: string;
  description?: string;
}
