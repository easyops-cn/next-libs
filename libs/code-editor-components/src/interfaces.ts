import { Annotation } from "brace";
import { FormItemWrapperProps } from "@next-libs/forms";

export interface Error {
  err: Annotation[];
  hasError: boolean;
}

export interface CodeEditorProps extends FormItemWrapperProps {
  placeholder?: string;
  value?: string;
  onChange?: (value: string) => void;
  onBlur?: () => void;
  onErrorChange?: (error: Error) => void;
  disabled?: boolean;
  theme?: string;
  mode: string;
  readOnly?: boolean;
  showLineNumbers?: boolean;
  maxLines?: number | "Infinity";
  minLines?: number;
  tabSize?: number;
  printMargin?: boolean;
  highlightActiveLine?: boolean;
  onValidate?: (error: Annotation[]) => void;
  showExportButton?: boolean;
  showCopyButton?: boolean;
  exportFileName?: string;
  jsonSchema?: Record<string, any>;
  validateJsonSchemaMode?: "warning" | "error";
  schemaRef?: string;
  enableLiveAutocompletion?: boolean;
  customCompleters?:
    | string[]
    | {
        caption?: string;
        value: string;
        meta?: string;
        score?: number;
      }[];
  loadYamlInJsonMode?: boolean;
  showPrintMargin?: boolean;
  celCompletersDisabled?: boolean;
}

export interface AceLanguageRules {
  [className: string]: AceLanguageRule[];
}

export interface AceLanguageRule {
  token?: string | string[] | AceTokenFunction;
  regex?: string | RegExp;
  next?: string;
  defaultToken?: string;
  consumeLineEnd?: boolean;
  onMatch?: (
    val?: string,
    state?: unknown,
    stack?: [string, number],
    line?: string
  ) => string;
}

export type AceTokenFunction = (value: string) => string;
