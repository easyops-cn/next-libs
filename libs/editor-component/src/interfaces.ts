import { Annotation } from "brace";

export interface Error {
  err: Annotation[];
  hasError: boolean;
}

export interface CodeEditorProps {
  placeholder?: string;
  value?: string;
  onChange?: (value: string) => void;
  onBlur?: () => void;
  onErrorChange?: (error: Error) => void;
  disabled?: boolean;
  theme?: string;
  mode: string;
  readOnly?: boolean;
  showLineNumbers?: boolean;
  maxLines?: number | "Infinity";
  minLines?: number;
  tabSize?: number;
  printMargin?: boolean;
  highlightActiveLine?: boolean;
  onValidate?: (error: Annotation[]) => void;
  showExportButton?: boolean;
  showCopyButton?: boolean;
  exportFileName?: string;
  jsonSchema?: Record<string, any>;
  schemaRef?: string;
  enableLiveAutocompletion?: boolean;
  customCompleters?:
    | string[]
    | {
        caption?: string;
        value: string;
        meta?: string;
        score?: number;
      }[];
}
