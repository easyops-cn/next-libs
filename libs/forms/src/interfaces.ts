import { WrappedFormUtils } from "@ant-design/compatible/lib/form/Form";
import { ColProps } from "antd/lib/col";
import React from "react";
import { MenuIcon, UseBrickConf } from "@next-core/brick-types";

export interface AbstractGeneralFormElement extends HTMLElement {
  formUtils: WrappedFormUtils;
  layout: string;
  noColon: boolean;
  labelCol: ColProps;
  wrapperCol: ColProps;
}

export type GeneralOption =
  | string
  | number
  | boolean
  | GeneralComplexOption
  | Record<string, any>;

export interface GeneralComplexOption<T = string | number | boolean> {
  label: string;
  value: T;
}

export interface LabelTooltipProps {
  content: string;
  icon: MenuIcon;
  popUpType?: "tooltip" | "popover";
  contentType?: "text" | "markdown";
  style?: React.CSSProperties;
  iconStyle?: React.CSSProperties;
}

export interface LabelBrick {
  useBrick: UseBrickConf;
}

export interface HelpBrickProps {
  useBrick: UseBrickConf;
  placement?: "right" | "bottom";
  containerStyle?: React.CSSProperties;
}
