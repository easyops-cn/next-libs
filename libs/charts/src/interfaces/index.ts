export * from "./format";
export * from "./panel";
