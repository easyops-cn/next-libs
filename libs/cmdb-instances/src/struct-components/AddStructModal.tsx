import React, { ChangeEvent } from "react";
import { DatePicker, Input, InputNumber, Modal, Radio, Select } from "antd";
import { Attribute, Structkey } from "./interfaces";
import { boolOptions } from "../model-attribute-form-control/ModelAttributeFormControl";
import styles from "./index.module.css";
import i18n from "i18next";
import { K, NS_LIBS_CMDB_INSTANCES } from "../i18n/constants";
import moment from "moment";
import { RadioChangeEvent } from "antd/lib/radio";
import { SelectValue } from "antd/lib/select";
import { computeDateFormat } from "../processors";
import _ from "lodash";
import { CodeEditor } from "@next-libs/code-editor-components";

export interface AddStructModalProps {
  structData?: any;
  attribute: Attribute;
  handleStoreFunction: Function; // eslint-disable-line
  handleCancelFunction: Function; // eslint-disable-line
  visible: boolean;
}
export interface AddStructModalState {
  structData?: any;
  // 格式是否匹配数组
  showError?: boolean[];
}

export class AddStructModal extends React.Component<
  AddStructModalProps,
  AddStructModalState
> {
  constructor(props: AddStructModalProps) {
    super(props);
    const { attribute } = props;
    this.state = {
      showError: new Array(attribute.value.struct_define.length).fill(false),
    };
  }
  componentDidUpdate(prevProps: AddStructModalProps) {
    if (prevProps.visible !== this.props.visible) {
      if (this.props.visible) {
        this.setState({ structData: this.props.structData || {} });
      }
    }
  }
  handleValueChange = (e: SelectValue, define: Structkey) => {
    const structData = this.state.structData;
    structData[define.id] = e;
    this.setState({ structData });
  };
  handleInputValueChange = (
    e:
      | RadioChangeEvent
      | ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>,
    define: any
  ) => {
    this.handleValueChange(e.target.value, define);
  };
  handleIpValueChange = (
    e: ChangeEvent<HTMLInputElement>,
    define: any,
    index: number
  ) => {
    const value = e.target.value;
    const { showError } = this.state;
    const matched =
      value === "" ||
      /^((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.){3}(25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)(\[^\\],;\s]{1,100}\]|)$/.test(
        value
      );
    showError[index] = !matched;
    this.setState({
      showError,
    });
    this.handleValueChange(value, define);
  };
  getEnumForm = (define: any, value: string) => {
    let enumForm;
    const { regex } = define;
    if (regex.length < 6) {
      const RadioGroup = Radio.Group;

      enumForm = (
        <RadioGroup
          onChange={(e) => this.handleInputValueChange(e, define)}
          defaultValue={value}
        >
          {regex.map((item: string) => (
            <Radio value={item} key={item}>
              {item}
            </Radio>
          ))}
        </RadioGroup>
      );
    } else {
      const Option = Select.Option;
      enumForm = (
        <Select
          {...(value !== null ? { defaultValue: value } : {})}
          style={{ width: "100%" }}
          onChange={(e: SelectValue) => this.handleValueChange(e, define)}
        >
          {regex.map((item: string) => (
            <Option value={item} key={item}>
              {item}
            </Option>
          ))}
        </Select>
      );
    }
    return enumForm;
  };
  validateJson = (err: any, index: number) => {
    const { showError } = this.state;
    const error = _.some(err, ["type", "error"]);
    showError[index] = error;
    this.setState({ showError });
  };
  getFormType = (define: Structkey, value: any, index: number) => {
    let formType;
    let defaultValue = value ? value[define.id] : null;
    //如果是json类型，数据有可能是字符串、数组或对象等，需要对数据处理
    if (define.type === "json") {
      defaultValue =
        _.isString(defaultValue) || !defaultValue
          ? defaultValue || ""
          : JSON.stringify(defaultValue, null, 2);
    }
    switch (define.type) {
      case "int": {
        formType = (
          <InputNumber
            defaultValue={defaultValue}
            style={{ width: "100%" }}
            precision={0}
            onChange={(e) => this.handleValueChange(e, define)}
          />
        );
        break;
      }
      case "enum": {
        formType = this.getEnumForm(define, defaultValue);
        break;
      }
      case "enums": {
        formType = (
          <Select
            mode="multiple"
            style={{ width: "100%" }}
            onChange={(e: SelectValue) => this.handleValueChange(e, define)}
            {...(defaultValue !== null ? { defaultValue: defaultValue } : {})}
          >
            {define.regex.map((item: string) => (
              <Select.Option value={item} key={item}>
                {item}
              </Select.Option>
            ))}
          </Select>
        );
        break;
      }
      case "bool": {
        formType = (
          <Radio.Group
            onChange={(e) => this.handleInputValueChange(e, define)}
            defaultValue={defaultValue}
          >
            {boolOptions.map((item) => (
              <Radio value={item.id} key={item.text}>
                {item.text}
              </Radio>
            ))}
          </Radio.Group>
        );
        break;
      }
      case "arr": {
        formType = (
          <Select
            mode="tags"
            style={{ width: "100%" }}
            {...(defaultValue !== null ? { defaultValue: defaultValue } : {})}
            onChange={(e: SelectValue) => this.handleValueChange(e, define)}
          />
        );
        break;
      }
      case "date":
      case "datetime": {
        const formater = computeDateFormat(define.type, defaultValue);
        formType = (
          <DatePicker
            defaultValue={formater.value}
            showTime={define.type === "datetime"}
            onChange={(e) =>
              this.handleValueChange(
                e.format(
                  define.type === "datetime"
                    ? "YYYY-MM-DD HH:mm:ss"
                    : "YYYY-MM-DD"
                ),
                define
              )
            }
            format={formater.format}
          />
        );
        break;
      }
      case "ip": {
        formType = (
          <div>
            <Input
              defaultValue={defaultValue}
              onChange={(e) => this.handleIpValueChange(e, define, index)}
            />
            <label
              style={{
                display: this.state.showError[index] ? "block" : "none",
                color: "#fc5043",
              }}
            >
              {i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.NOT_MEET_REGEX}`)}
            </label>
          </div>
        );
        break;
      }
      case "float": {
        formType = (
          <InputNumber
            defaultValue={defaultValue}
            style={{ width: "100%" }}
            onChange={(e) => this.handleValueChange(e, define)}
          />
        );
        break;
      }
      case "json": {
        formType = (
          <div>
            <CodeEditor
              value={defaultValue}
              mode={"json"}
              maxLines={"Infinity"}
              highlightActiveLine={true}
              onChange={(e: any) => this.handleValueChange(e, define)}
              minLines={3}
              showLineNumbers={true}
              showPrintMargin={false}
              onValidate={(err: any) => this.validateJson(err, index)}
            />
            <label
              style={{
                display: this.state.showError[index] ? "block" : "none",
                color: "#fc5043",
              }}
            >
              {i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.NOT_MEET_JSON}`)}
            </label>
          </div>
        );
        break;
      }
      default: {
        formType = (
          <Input
            defaultValue={defaultValue}
            onChange={(e) => this.handleInputValueChange(e, define)}
          />
        );
      }
    }
    return formType;
  };

  // 保存修改
  handleStore = () => {
    this.props.handleStoreFunction(this.state.structData);
  };
  handleCloseModal = () => {
    this.props.handleCancelFunction();
  };
  render() {
    const { attribute, visible, structData } = this.props;
    const structDefine = attribute.value.struct_define;
    return (
      <Modal
        visible={visible}
        onOk={this.handleStore}
        onCancel={this.handleCloseModal}
        okButtonProps={{
          disabled: this.state.showError.includes(true),
          id: "okBtn",
        }}
        okText={i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.CONFIRM}`)}
        cancelText={i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.CANCEL}`)}
        cancelButtonProps={{ id: "cancelBtn" }}
        destroyOnClose={true}
      >
        {structDefine.map((define, index) => {
          return (
            <div className={styles.formItem} key={define.id}>
              <label className={styles.formLabel}>{define.name}：</label>
              {this.getFormType(define, structData, index)}
            </div>
          );
        })}
      </Modal>
    );
  }
}
