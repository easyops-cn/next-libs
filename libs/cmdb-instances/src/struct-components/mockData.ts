import { Attribute } from "./interfaces";
export const structData = {
  arr: ["abc", "def"],
  str: "string",
  int: 1000,
  ip: "0.0.0.0",
  enum: "1",
  date: "2019-05-24",
  datetime: "2019-05-24 12:00:00",
  enums: ["1", "3"],
  float: 100.1,
  json: "666",
  json1: { test: "666" },
};
export const structData2 = {
  arr: ["ppp"],
  str: "string2",
  int: 100,
  ip: "10.0.0.0",
  enum: "5",
  date: "2019-05-23",
  datetime: "2019-05-24 23:00:00",
  enums: ["1", "3", "5"],
  float: 100.1,
};
export const structList = [structData, structData2];
export const attribute: Attribute = {
  name: "结构体",
  id: "struct",
  value: {
    type: "structs",
    struct_define: [
      {
        id: "arr",
        name: "数组",
        type: "arr",
      },
      {
        id: "str",
        name: "字符串",
        type: "str",
      },
      {
        id: "int",
        name: "数组",
        type: "int",
      },
      {
        id: "ip",
        name: "IP地址",
        type: "ip",
      },
      {
        id: "enum",
        name: "枚举",
        type: "enum",
        regex: ["1", "3", "5", "7", "9"],
      },
      {
        id: "date",
        name: "日期",
        type: "date",
      },
      {
        id: "datetime",
        name: "时间",
        type: "datetime",
      },
      {
        id: "enums",
        name: "枚举多选",
        type: "enums",
        regex: ["1", "3", "5", "7", "9"],
      },
      {
        id: "float",
        name: "浮点型",
        type: "float",
      },
      {
        id: "json",
        name: "json",
        type: "json",
      },
      {
        id: "json1",
        name: "json1",
        type: "json",
      },
    ],
  },
};
