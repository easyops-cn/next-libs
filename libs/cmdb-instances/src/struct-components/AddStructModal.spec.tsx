import React, { ChangeEvent } from "react";
import { shallow } from "enzyme";
import { AddStructModal, AddStructModalProps } from "./AddStructModal";
import { DatePicker, Input, InputNumber, Radio, Select } from "antd";
import moment from "moment";
import { attribute, structData } from "./mockData";
import { RadioChangeEvent } from "antd/lib/radio";
import { Attribute } from "./interfaces";

describe("AddStructModal", () => {
  const handleStoreFunction = jest.fn();
  const handleCancelFunction = jest.fn();
  const props = {
    structData,
    attribute,
    handleStoreFunction,
    handleCancelFunction,
    visible: false,
  };
  const wrapper = shallow<AddStructModalProps>(<AddStructModal {...props} />);
  const instance = wrapper.instance() as AddStructModal;
  it("should init addStructModal", () => {
    expect(wrapper).toBeTruthy();
    expect(instance.state.showError).toEqual(new Array(8).fill(false));
  });
  it("should call the store function", () => {
    const spy = jest.spyOn(props, "handleStoreFunction");

    wrapper.setProps({ visible: true });
    instance.handleStore();
    expect(spy).toBeCalledWith(structData);
  });
  it("should call the cancel function", () => {
    const spy = jest.spyOn(props, "handleCancelFunction");
    instance.handleCloseModal();
    expect(spy).toBeCalled();
  });
  it("should call the handleValueChange when change the int value", () => {
    wrapper.find(InputNumber).prop("onChange")(100);
    expect(instance.state.structData["int"]).toEqual(100);
  });
  it("should call the handleValueChange when change the string value", () => {
    wrapper.find(Input).at(0).prop("onChange")({
      target: { value: "newString" },
    } as ChangeEvent<HTMLInputElement>);
    expect(instance.state.structData["str"]).toEqual("newString");
  });

  it("should call the handleIpValueChange when change the ip value", () => {
    wrapper.find(Input).at(1).prop("onChange")({
      target: { value: "1.1.1.1" },
    } as ChangeEvent<HTMLInputElement>);
    expect(instance.state.structData["ip"]).toEqual("1.1.1.1");
  });
  it("should show the error message when give a invalid ip", () => {
    wrapper.find(Input).at(1).prop("onChange")({
      target: { value: "1.1.1" },
    } as ChangeEvent<HTMLInputElement>);
    expect(instance.state.showError[3]).toEqual(true);
  });
  it("should render radios group when enum's count is smaller than 6", () => {
    expect(wrapper.find(Radio.Group).length).toEqual(1);
  });
  it("should change the enum value", () => {
    wrapper.find(Radio.Group).prop("onChange")({
      target: { value: "3" },
    } as RadioChangeEvent);
    expect(instance.state.structData["enum"]).toEqual("3");
  });
  it("should change the arr value", () => {
    wrapper.find(Select).at(0).prop("onChange")(
      ["abc", "xyz"],
      expect.anything()
    );
    expect(instance.state.structData["arr"]).toEqual(["abc", "xyz"]);
  });
  it("should change the date value", () => {
    wrapper.find(DatePicker).at(0).prop("onChange")(
      moment("2019-01-01"),
      expect.anything()
    );
    expect(instance.state.structData["date"]).toEqual("2019-01-01");
  });
  it("should change the datetime value", () => {
    wrapper.find(DatePicker).at(1).prop("onChange")(
      moment("2019-01-01 23:59:59"),
      expect.anything()
    );
    expect(instance.state.structData["datetime"]).toEqual(
      "2019-01-01 23:59:59"
    );
  });
  it("should change the enums value", () => {
    wrapper.find(Select).at(1).prop("onChange")(
      ["3", "5", "7"],
      expect.anything()
    );
    expect(instance.state.structData["enums"]).toEqual(["3", "5", "7"]);
  });
  it("should render radio group", () => {
    const define = {
      id: "enum",
      name: "枚举",
      type: "enum",
      regex: ["1", "3", "5", "7", "9"],
    };
    const radioWrapper = shallow(instance.getEnumForm(define, "1"));
    expect(radioWrapper).toBeTruthy();
  });
  it("should render Select when enum", () => {
    const define = {
      id: "enum",
      name: "枚举",
      type: "enum",
      regex: ["1", "3", "5", "7", "9", "11"],
    };
    const selectWrapper = shallow(instance.getEnumForm(define, "1"));
    expect(selectWrapper).toBeTruthy();
  });

  it("should render Radio base on bool type", () => {
    const attribute: Attribute = {
      name: "结构体",
      id: "struct",
      value: {
        type: "structs",
        struct_define: [
          {
            id: "isShow",
            name: "显示",
            type: "bool",
          },
        ],
      },
    };

    const wrapper2 = shallow<AddStructModalProps>(
      <AddStructModal {...props} attribute={attribute} />
    );
    wrapper2.setProps({
      visible: true,
    });

    wrapper2.find(Radio.Group).invoke("onChange")({
      target: { value: true },
    } as RadioChangeEvent);
    expect(wrapper2.find(Radio.Group).prop("defaultValue")).toEqual(true);
  });
});
