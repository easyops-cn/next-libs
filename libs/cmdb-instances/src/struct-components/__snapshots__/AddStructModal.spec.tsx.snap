// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStructModal should render radio group 1`] = `
<ContextProvider
  value={
    Object {
      "disabled": undefined,
      "name": undefined,
      "onChange": [Function],
      "value": "1",
    }
  }
>
  <div
    className="ant-radio-group ant-radio-group-undefined"
  >
    <Radio
      key="1"
      type="radio"
      value="1"
    >
      1
    </Radio>
    <Radio
      key="3"
      type="radio"
      value="3"
    >
      3
    </Radio>
    <Radio
      key="5"
      type="radio"
      value="5"
    >
      5
    </Radio>
    <Radio
      key="7"
      type="radio"
      value="7"
    >
      7
    </Radio>
    <Radio
      key="9"
      type="radio"
      value="9"
    >
      9
    </Radio>
  </div>
</ContextProvider>
`;
